File: lessons/07-langchain-rag/python/langchain_rag.py
"""Lesson 7 - Rebuild RAG with LangChain, and count what it cost.

Runs the Lesson 1 pipeline and the LangChain pipeline over the same corpus, with
the same system prompt, and prints:

  1. per question, what each side retrieved and cited - do they ground the same way?
  2. component by component, which hand-rolled file each LangChain piece replaced
  3. what the framework cost: your lines of code against your dependency surface

The default run is deterministic and offline: it compares *grounding*, not
generated text. Retrieval and the rendered prompt are reproducible; a model's
prose is not, and comparing what both pipelines feed the model is the sharper
test anyway.

Run:
  python langchain_rag.py                    # the comparison (offline, no model)
  python langchain_rag.py --measure          # re-derive the dependency numbers here
  python langchain_rag.py ask "question"     # real answer, via your chat model adapter
  python langchain_rag.py ask --native "..." # real answer, via framework-native ChatOllama

PRODUCTION (see the lesson README, "From demo to production"):
- the comparison is the teaching artefact, not the product; in a real system you
  pick one pipeline. What carries over is the citation contract and the habit of
  measuring the dependency you are about to take.
"""

import ast
import io
import json
import sys
import time
import tokenize
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

import handrolled_pipeline as handrolled  # noqa: E402

LESSON_DIR = Path(__file__).resolve().parent.parent
ROOT = handrolled.ROOT
CONFIG_FILE = LESSON_DIR / "data" / "questions.json"

INSTALL_HINT = "pip install -r lessons/07-langchain-rag/requirements.txt"

# Which LangChain component stands in for which Lesson 1 file.
COMPONENTS = [
    ("load", "localrag/extract.py", "Document(...)"),
    ("split", "localrag/chunk.py", "RecursiveCharacterTextSplitter"),
    ("index", "localrag/store.py", "InMemoryVectorStore"),
    ("retrieve", "localrag/retriever.py", "BaseRetriever subclass (yours)"),
    ("prompt", "localrag/prompts.py", "ChatPromptTemplate"),
    ("provider", "localrag/providers/__init__.py", "SimpleChatModel subclass (yours)"),
    ("chain", "localrag/engine.py", "LCEL  ( | )"),
]

# The lesson's own LangChain code - the part you still write.
LC_FILES = ["python/lc_pipeline.py", "python/lc_provider.py"]

# Measured with `pip download`; re-derive the installed counts with --measure.
# Reproduce:  pip download --no-deps -d /tmp/x langchain-core langchain-text-splitters
DEPS = {
    "handrolled": {"direct": 8, "transitive": 8, "size_mb": 9},
    "langchain": {"direct": 10, "transitive": 31, "size_mb": 13},
}


def code_lines(path: Path) -> int:
    """Lines that are actually code: no blanks, no comments, no docstrings.

    Both sides are counted the same way, so a teaching file full of explanation
    is not unfairly penalised against a terse one.
    """
    source = path.read_text(encoding="utf-8")
    doc_lines = set()
    tree = ast.parse(source)
    for node in ast.walk(tree):
        if isinstance(node, (ast.Module, ast.ClassDef, ast.FunctionDef, ast.AsyncFunctionDef)):
            body = getattr(node, "body", None)
            if body and isinstance(body[0], ast.Expr) and isinstance(body[0].value, ast.Constant):
                if isinstance(body[0].value.value, str):
                    doc_lines.update(range(body[0].lineno, body[0].end_lineno + 1))
    counted = set()
    for tok in tokenize.generate_tokens(io.StringIO(source).readline):
        if tok.type in (tokenize.COMMENT, tokenize.NL, tokenize.NEWLINE, tokenize.INDENT,
                        tokenize.DEDENT, tokenize.ENDMARKER, tokenize.ENCODING):
            continue
        if tok.start[0] not in doc_lines:
            counted.add(tok.start[0])
    return len(counted)


def total_lines(rel_paths) -> int:
    return sum(code_lines(ROOT / rel) for rel in rel_paths)


def load_settings() -> dict:
    with open(CONFIG_FILE, encoding="utf-8") as fh:
        return json.load(fh)


def import_langchain():
    """Return the LangChain pipeline module, or None when it is not installed."""
    try:
        import lc_pipeline
        return lc_pipeline
    except ImportError:
        return None


def print_questions(settings, lc) -> None:
    size, overlap, k = settings["chunk_size"], settings["chunk_overlap"], settings["top_k"]
    hand_chunks = handrolled.split(handrolled.load(), size, overlap)
    lc_chunks = lc.split(lc.load(), size, overlap) if lc else []
    docs = len({c["source"] for c in hand_chunks})

    print(f"Rebuild RAG with LangChain  -  {docs} documents, same corpus, same system prompt")
    print(f"chunked at size={size} overlap={overlap}, retrieving top {k}")
    print()
    print(f"  hand-rolled  {len(hand_chunks):3} chunks   (chunk.py: collapse whitespace, "
          f"break on sentences)")
    if lc:
        print(f"  langchain    {len(lc_chunks):3} chunks   (RecursiveCharacterTextSplitter: "
              f"keep text, split on separators)")
    print()

    retriever = lc.bm25_retriever(lc_chunks, k) if lc else None
    agreements = 0
    for i, question in enumerate(settings["questions"], start=1):
        hand_sources = handrolled.sources(handrolled.retrieve(hand_chunks, question, k))
        print(f"Q{i}  {question}")
        print(f"    hand-rolled   sources: {' . '.join(hand_sources)}")
        if not lc:
            print("    langchain     not installed")
            print()
            continue
        lc_sources = lc.sources(retriever.invoke(question))
        print(f"    langchain     sources: {' . '.join(lc_sources)}")
        if hand_sources == lc_sources:
            agreements += 1
            print("    GROUNDING AGREES  -  same sources, same order")
        else:
            only_lc = [s for s in lc_sources if s not in hand_sources]
            only_hand = [s for s in hand_sources if s not in lc_sources]
            extra = []
            if only_lc:
                extra.append("langchain also cites " + ", ".join(only_lc))
            if only_hand:
                extra.append("hand-rolled also cites " + ", ".join(only_hand))
            print(f"    GROUNDING DIFFERS  -  {'; '.join(extra)}")
        print()
    if lc:
        total = len(settings["questions"])
        print(f"    {agreements}/{total} questions grounded identically. "
              f"Different chunk boundaries, mostly the same evidence.")
        print()


def print_components(lc) -> None:
    print("Component by component")
    print(f"  {'step':9} {'hand-rolled (Lesson 1)':36} {'LangChain (Lesson 7)'}")
    print("  " + "-" * 79)
    hand_total = 0
    for step, rel, component in COMPONENTS:
        loc = code_lines(ROOT / rel)
        hand_total += loc
        shown = component if lc else "not installed"
        print(f"  {step:9} {rel:31}{loc:>4}L  {shown}")
    print("  " + "-" * 79)

    if not lc:
        print()
        print("LangChain is not installed, so only the hand-rolled side ran.")
        print(f"Install it to see the comparison:  {INSTALL_HINT}")
        return

    lc_total = total_lines([f"lessons/07-langchain-rag/{p}" for p in LC_FILES])
    print()
    print("What it cost")
    print(f"  {'':22} {'hand-rolled':>14} {'LangChain':>14}")
    print(f"  {'code you maintain':22} {str(hand_total) + ' lines':>14} "
          f"{str(lc_total) + ' lines':>14}")
    print(f"  {'direct dependencies':22} {DEPS['handrolled']['direct']:>14} "
          f"{DEPS['langchain']['direct']:>14}")
    print(f"  {'transitive packages':22} {DEPS['handrolled']['transitive']:>14} "
          f"{DEPS['langchain']['transitive']:>14}")
    print(f"  {'install size':22} {'~' + str(DEPS['handrolled']['size_mb']) + ' MB':>14} "
          f"{'~' + str(DEPS['langchain']['size_mb']) + ' MB':>14}")
    print()
    print(f"  {hand_total} lines you can read against {lc_total} lines you still write.")
    print("  The rest moved into 31 packages you did not read. That is the trade.")


def cmd_demo() -> int:
    settings = load_settings()
    lc = import_langchain()
    print_questions(settings, lc)
    print_components(lc)
    return 0


def cmd_measure() -> int:
    """Re-derive the dependency and start-up numbers on THIS machine."""
    from importlib.metadata import distributions
    lc = import_langchain()
    print("Measured here, now")
    print(f"  installed distributions: {len(list(distributions()))}")
    start = time.perf_counter()
    __import__("localrag.engine")
    print(f"  import localrag.engine:  {time.perf_counter() - start:.3f}s")
    if lc:
        start = time.perf_counter()
        __import__("langchain_core.runnables")
        print(f"  import langchain_core:   {time.perf_counter() - start:.3f}s")
    else:
        print("  langchain_core:          not installed")
    print()
    print("Reproduce the package counts with:")
    print("  pip download --no-deps -d /tmp/x langchain-core langchain-text-splitters")
    return 0


def cmd_ask(question: str, native: bool, arm: str) -> int:
    lc = import_langchain()
    if lc is None:
        print(f"LangChain is not installed. Install it with:\n  {INSTALL_HINT}")
        return 1
    settings = load_settings()
    config = handrolled.lesson_config()
    chunks = lc.split(lc.load(), settings["chunk_size"], settings["chunk_overlap"])

    if arm == "embed":
        model = config.ollama_embed_model
        retriever = (lc.native_ollama_retriever(chunks, settings["top_k"], model) if native
                     else lc.embedding_retriever(chunks, settings["top_k"], config))
    else:
        retriever = lc.bm25_retriever(chunks, settings["top_k"])

    hits = retriever.invoke(question)
    print(f"Q  {question}")
    print(f"   sources: {' . '.join(lc.sources(hits))}")
    print()

    if native:
        from langchain_ollama import ChatOllama
        llm = ChatOllama(model=config.ollama_model)
        label = f"ChatOllama({config.ollama_model}) - framework-native"
    else:
        from lc_provider import LocalRagChatModel
        llm = LocalRagChatModel(config=config, provider_name=config.provider)
        label = f"LocalRagChatModel({config.provider}) - your adapter"

    print(f"   model: {label}")
    print()
    print(lc.build_chain(retriever, llm).invoke(question))
    return 0


def main(argv=None) -> int:
    argv = list(sys.argv[1:] if argv is None else argv)
    native = "--native" in argv
    argv = [a for a in argv if a != "--native"]
    arm = "bm25"
    if "--arm" in argv:
        i = argv.index("--arm")
        arm = argv[i + 1] if i + 1 < len(argv) else "bm25"
        del argv[i:i + 2]

    if argv and argv[0] == "--measure":
        return cmd_measure()
    if argv and argv[0] == "ask":
        question = " ".join(argv[1:]).strip()
        if not question:
            print('Usage: python langchain_rag.py ask "your question"')
            return 2
        return cmd_ask(question, native, arm)
    return cmd_demo()


if __name__ == "__main__":
    raise SystemExit(main())

File: lessons/07-langchain-rag/node/langchain_rag.mjs
// Lesson 7 - Rebuild RAG with LangChain, Node.js port.
//
// LangChain ships two official SDKs: Python and JavaScript. This file is the
// second one, and it exists to make a point the Python side cannot: the same
// components, the same LCEL-style composition, a different runtime - and a
// noticeably different dependency bill.
//
// The hand-rolled half is a direct port of Lesson 1's Python (extract -> chunk ->
// BM25 -> cite), so the retrieval half of the output matches the Python run byte
// for byte. The scorecard half does not, and must not: 18 added packages / 9 MB on
// PyPI against 12 added packages / 48 MB on npm is the honest result.
//
// Run:  node langchain_rag.mjs        (from lessons/07-langchain-rag/node)

import { readFileSync, readdirSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { Document } from "@langchain/core/documents";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const LESSON_DIR = path.resolve(HERE, "..");
const CORPUS_DIR = path.join(LESSON_DIR, "data", "corpus");
const SETTINGS = JSON.parse(readFileSync(path.join(LESSON_DIR, "data", "questions.json"), "utf8"));

// --------------------------------------------------------------- hand-rolled (Lesson 1)

// localrag/retriever.py:_tokenize
function tokenize(text) {
  return text.toLowerCase().match(/[a-z0-9]+/g) ?? [];
}

// localrag/extract.py - one page per markdown file, sorted for a stable index.
function handLoad() {
  return readdirSync(CORPUS_DIR)
    .filter((name) => name.endsWith(".md"))
    .sort()
    .map((name) => ({
      source: name,
      page_number: 1,
      text: readFileSync(path.join(CORPUS_DIR, name), "utf8").trim(),
    }));
}

// localrag/chunk.py:_split_text - collapse whitespace first, then break on a
// sentence boundary past the halfway mark.
function splitText(text, size, overlap) {
  const normalized = text.split(/\s+/).filter(Boolean).join(" ");
  if (normalized.length <= size) return normalized ? [normalized] : [];

  const chunks = [];
  let start = 0;
  const n = normalized.length;
  while (start < n) {
    let end = Math.min(start + size, n);
    if (end < n) {
      const window = normalized.slice(start, end);
      for (const sep of [". ", "! ", "? ", "\n", " "]) {
        const pos = window.lastIndexOf(sep);
        if (pos > Math.floor(size / 2)) {
          end = start + pos + sep.length;
          break;
        }
      }
    }
    const piece = normalized.slice(start, end).trim();
    if (piece) chunks.push(piece);
    if (end >= n) break;
    start = Math.max(end - overlap, start + 1);
  }
  return chunks;
}

function handSplit(pages, size, overlap) {
  const chunks = [];
  let index = 0;
  for (const page of pages) {
    for (const piece of splitText(page.text, size, overlap)) {
      chunks.push({ source: page.source, page_number: page.page_number, chunk_index: index, text: piece });
      index += 1;
    }
  }
  return chunks;
}

// rank_bm25.BM25Okapi, ported exactly - including the epsilon floor it applies to
// terms that appear in more than half the corpus. Same numbers, same ordering.
class BM25Okapi {
  constructor(corpus, k1 = 1.5, b = 0.75, epsilon = 0.25) {
    this.k1 = k1;
    this.b = b;
    this.corpusSize = corpus.length;
    this.docFreqs = [];
    this.docLen = [];
    this.idf = new Map();

    const nd = new Map();
    let numDoc = 0;
    for (const document of corpus) {
      this.docLen.push(document.length);
      numDoc += document.length;
      const frequencies = new Map();
      for (const word of document) frequencies.set(word, (frequencies.get(word) ?? 0) + 1);
      this.docFreqs.push(frequencies);
      for (const word of frequencies.keys()) nd.set(word, (nd.get(word) ?? 0) + 1);
    }
    this.avgdl = numDoc / this.corpusSize;

    let idfSum = 0;
    const negative = [];
    for (const [word, freq] of nd) {
      const idf = Math.log(this.corpusSize - freq + 0.5) - Math.log(freq + 0.5);
      this.idf.set(word, idf);
      idfSum += idf;
      if (idf < 0) negative.push(word);
    }
    const eps = epsilon * (idfSum / this.idf.size);
    for (const word of negative) this.idf.set(word, eps);
  }

  getScores(query) {
    const scores = new Array(this.corpusSize).fill(0);
    for (const q of query) {
      const idf = this.idf.get(q) ?? 0;
      for (let i = 0; i < this.corpusSize; i += 1) {
        const freq = this.docFreqs[i].get(q) ?? 0;
        const denom = freq + this.k1 * (1 - this.b + (this.b * this.docLen[i]) / this.avgdl);
        scores[i] += (idf * (freq * (this.k1 + 1))) / denom;
      }
    }
    return scores;
  }
}

// One index per corpus, built on first use and reused for every later query -
// the same contract as Lesson 1's Bm25Retriever.__init__ and as the Python arms.
// Keyed on the array itself, so the two corpora here never share an index.
const bm25Indexes = new WeakMap();

function indexFor(items, textOf) {
  let bm25 = bm25Indexes.get(items);
  if (!bm25) {
    bm25 = new BM25Okapi(items.map((it) => tokenize(textOf(it))));
    bm25Indexes.set(items, bm25);
  }
  return bm25;
}

// localrag/retriever.py:Bm25Retriever.search - top-k, dropping the tail that
// scores at or below zero once there is a real signal.
function bm25Search(items, textOf, query, k) {
  if (!items.length || k <= 0) return [];
  const scores = indexFor(items, textOf).getScores(tokenize(query));
  const ranked = items
    .map((_, i) => i)
    .sort((a, b) => scores[b] - scores[a])
    .slice(0, k);
  if (scores[ranked[0]] <= 0) return ranked.map((i) => items[i]);
  return ranked.filter((i) => scores[i] > 0).map((i) => items[i]);
}

// localrag/engine.py:dedup_sources
function citations(pairs) {
  const seen = [];
  for (const [source, page] of pairs) {
    const tag = `${source}:${page}`;
    if (!seen.includes(tag)) seen.push(tag);
  }
  return seen;
}

// --------------------------------------------------------------- LangChain

function lcLoad() {
  return readdirSync(CORPUS_DIR)
    .filter((name) => name.endsWith(".md"))
    .sort()
    .map(
      (name) =>
        new Document({
          pageContent: readFileSync(path.join(CORPUS_DIR, name), "utf8").trim(),
          metadata: { source: name, page: 1 },
        }),
    );
}

async function lcSplit(docs, size, overlap) {
  const splitter = new RecursiveCharacterTextSplitter({ chunkSize: size, chunkOverlap: overlap });
  return splitter.splitDocuments(docs);
}

// --------------------------------------------------------------- report

async function main() {
  const { chunk_size: size, chunk_overlap: overlap, top_k: k, questions } = SETTINGS;

  const handChunks = handSplit(handLoad(), size, overlap);
  const lcChunks = await lcSplit(lcLoad(), size, overlap);
  const docs = new Set(handChunks.map((c) => c.source)).size;

  const out = [];
  out.push(`Rebuild RAG with LangChain  -  ${docs} documents, same corpus, same system prompt`);
  out.push(`chunked at size=${size} overlap=${overlap}, retrieving top ${k}`);
  out.push("");
  out.push(`  hand-rolled  ${String(handChunks.length).padStart(3)} chunks   (chunk.py: collapse whitespace, break on sentences)`);
  out.push(`  langchain    ${String(lcChunks.length).padStart(3)} chunks   (RecursiveCharacterTextSplitter: keep text, split on separators)`);
  out.push("");

  let agreements = 0;
  questions.forEach((question, i) => {
    const handHits = bm25Search(handChunks, (c) => c.text, question, k);
    const handSources = citations(handHits.map((c) => [c.source, c.page_number]));
    const lcHits = bm25Search(lcChunks, (d) => d.pageContent, question, k);
    const lcSources = citations(lcHits.map((d) => [d.metadata.source, d.metadata.page]));

    out.push(`Q${i + 1}  ${question}`);
    out.push(`    hand-rolled   sources: ${handSources.join(" . ")}`);
    out.push(`    langchain     sources: ${lcSources.join(" . ")}`);
    if (handSources.join("|") === lcSources.join("|")) {
      agreements += 1;
      out.push("    GROUNDING AGREES  -  same sources, same order");
    } else {
      const onlyLc = lcSources.filter((s) => !handSources.includes(s));
      const onlyHand = handSources.filter((s) => !lcSources.includes(s));
      const extra = [];
      if (onlyLc.length) extra.push(`langchain also cites ${onlyLc.join(", ")}`);
      if (onlyHand.length) extra.push(`hand-rolled also cites ${onlyHand.join(", ")}`);
      out.push(`    GROUNDING DIFFERS  -  ${extra.join("; ")}`);
    }
    out.push("");
  });
  out.push(`    ${agreements}/${questions.length} questions grounded identically. Different chunk boundaries, mostly the same evidence.`);
  out.push("");
  console.log(out.join("\n"));

  console.log("What it cost, on npm");
  console.log("  @langchain/core + @langchain/textsplitters");
  console.log("  installed packages   12   (from a baseline of none at all)");
  console.log("  install size         ~48 MB");
  console.log("");
  console.log("  The Python run adds 18 packages and ~9 MB for the same pipeline,");
  console.log("  on top of a course baseline that is already 49 packages.");
  console.log("  Same framework, same components, a different bill. Measure the one you ship.");
}

// Surface failures instead of leaving an unhandled rejection: without this a bad
// data/questions.json or a missing corpus file could end the process with a stack
// trace and a zero exit status, which `./run` would read as success.
main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
